### PR TITLE
Fetch `authenticatedUser` server-side in Next.js

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR adds `getInitialProps` to `_app.page.ts` which fetches the `authenticatedUser` server-side on _every_ page-load in the HASH App. This allows the `authenticatedUser` to be available on first-render of the application (when the user is logged in).

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203363157432081/1203451531168818/f) _(internal)_

## 🔍 What does this change?

- replaces the existing `useAuthenticatedUser` with a simplified version of the hook, that makes use of react context to minimise data fetching
- fixes some routing logic to allow pages to be publicly accessible in readonly mode
- updates the `oryKratosClient` to be work on both the server and the client

## 📜 Does this require a change to the docs?

Not that I'm aware of.

## 🛡 What tests cover this?

None, playwright is not working. The app was tested manually.

## ❓ How to test this?

1. Run the HASH App, and login.
2. Ensure the app is working as previously, and that redirects are happening before the page is rendered.

### CODE REVIEW

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts [hunk 1]
```diff
@@ -41,23 +44,27 @@ export const useUpdateAuthenticatedUser = () => {
           return { updatedAuthenticatedUser: authenticatedUser };
         }
 
-        const [
-          {
-            data: { me: latestMeSubgraph },
-          },
-        ] = await refetch();
+        const { data: latestUserEntitySubgraph } = await getEntity({
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:47]: Do you need the default depths here? I'm guessing you actually want depths of 0?

File: packages/hash/frontend/src/pages/_app.page.tsx [hunk 2]
```diff
@@ -131,18 +129,110 @@ const App: FunctionComponent<AppProps> = ({
...
+};
+
+// The list of page pathnames that should be accessible whether or not the user is authenticated
+const publiclyAccessiblePagePathnames = [
+  "/[account-slug]/[page-slug]",
+  "/login",
+  "/signup",
+];
+
+AppWithTypeSystemContextProvider.getInitialProps = async (appContext) => {
```
Comment (benwerner01) [inline packages/hash/frontend/src/pages/_app.page.tsx:184]: Note to reviewers that [it's not possible to use `getServerSideProps` in `_app`](https://nextjs.org/docs/advanced-features/custom-app#caveats)

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts (hunk 1 above)
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Whilst testing the signup flow I realised the `getEntity` BP method doesn't work for authenticated users that haven't completed signup. So have ended up replacing this with the `me` query as it was previously in https://github.com/hashintel/hash/pull/1637/commits/a9e2f792e65eccf98954ecef630a89c3b1cfb7bb

File: packages/hash/frontend/src/pages/_app.page.tsx [hunk 3]
```diff
@@ -57,35 +62,27 @@ const App: FunctionComponent<AppProps> = ({
...
+        authenticatedUser &&
+        sentryUser?.id !== authenticatedUser.entityEditionId.baseId
+      ) {
+        const primaryEmail = authenticatedUser.emails.find(
+          (email) => email.primary,
+        );
+        Sentry.setUser({
+          id: authenticatedUser.entityEditionId.baseId,
+          email: primaryEmail?.address,
+        });
```
Comment (yusufkinatas) [inline packages/hash/frontend/src/pages/_app.page.tsx:82]: Can this be in a separate folder to keep `_app` easier to read?

File: packages/hash/frontend/src/pages/_app.page.tsx [hunk 4]
```diff
@@ -131,18 +128,112 @@ const App: FunctionComponent<AppProps> = ({
...
+    if (!res.writeHead) {
+      // `res.writeHead` is not a function whilst building next, so return instead.
+      return;
+    }
+    res.writeHead(307, { Location: location });
+    res.end();
+  } else {
+    void router.push(location);
+  }
+};
```
Comment (yusufkinatas) [inline packages/hash/frontend/src/pages/_app.page.tsx:176]: Same, can this be in a separate folder to keep this file easier to read?

File: packages/hash/frontend/src/pages/_app.page.tsx [hunk 5]
```diff
@@ -131,18 +128,112 @@ const App: FunctionComponent<AppProps> = ({
...
+
+// The list of page pathnames that should be accessible whether or not the user is authenticated
+const publiclyAccessiblePagePathnames = [
+  "/[account-slug]/[page-slug]",
+  "/login",
+  "/signup",
+];
+
+AppWithTypeSystemContextProvider.getInitialProps = async (appContext) => {
```
Comment (yusufkinatas) [inline packages/hash/frontend/src/pages/_app.page.tsx:185]: Same for this, we could move `App`'s getInitialProps function, or just some parts of it to a separate file to make this file more readable, it looks pretty crowded right now with 200+ lines, hard to follow.

**Review by yusufkinatas — APPROVED** [review]
Great work 💪
Tested on my local and it works as expected.

Looks good to me, but noted my concerns about readability / ease to read of `_app.page.tsx` file above. Let me know what do you think.

File: packages/hash/frontend/src/pages/_app.page.tsx (hunk 4 above)
Comment (benwerner01) [reply packages/hash/frontend/src/pages/_app.page.tsx]: I've moved this function in to a separate file as part of https://github.com/hashintel/hash/pull/1637/commits/22b40db8e145c4776c7a95c1a1099ed6010bc4f6

File: packages/hash/frontend/src/pages/_app.page.tsx (hunk 3 above)
Comment (benwerner01) [reply packages/hash/frontend/src/pages/_app.page.tsx]: I've encapsulated this in a function and moved it into a separate file as part of https://github.com/hashintel/hash/pull/1637/commits/22b40db8e145c4776c7a95c1a1099ed6010bc4f6

**Review by yusufkinatas — APPROVED** [review]
@benwerner01 I think it's better now, thanks for the quick response. Good to go for me!
